### ISSUE TITLE
Add descriptions to function returns / table columns?

### ISSUE THREAD

**Issue by Am3ra** [jldbc/pybaseball#97]
I've been developing some projects, and one of the main pet peeves (let me be clear, it isn't major) is not having a good description of the tables/of the columns that they have.

I wonder if other people would appreciate having descriptions of the tables centralized in the documentation of this repo?

- Pros: 
   - Easier use of the functions, knowing exactly what they return (What shape and what columns, maybe even what columns mean)
- Cons: It's possible that the various databases chage their shape (low risk, they are pretty stable)

Basically more documentation for function outputs.

What do you guys think?

**Comment by schorrm** [comment]
Statcast has a documentation file, I think.
Columns is pretty easy, `df.columns` gives it to you. If we're talking stuff beyond that, that's okay, in external (Markdown) description. Not as a signature or docstring.

**Comment by bdilday** [comment]
+1 to defining the schema of data frames somewhere. I'd suggest doing it in code rather than text files though. maybe use something like `pandera`, which defines schemas for pandas data frames (and does validation)? https://github.com/pandera-dev/pandera

**Comment by schorrm** [comment]
I'm not sure I really see the payoff there?

**Comment by Am3ra** [comment]
@schorrm Why not as a docstring? I don't really mind how we represent the function returns as long as there is some more thorough description of each return somewhere.

As you say, adding a markdown description seems an easy way to achieve this, and I think just adding the column descriptions as a table to the documentation should be fairly easy. 

Regarding @bdilday , I don't see how pandera solves the problem of not completely understanding the values that the functions return, or what they represent.

**Comment by schorrm** [comment]
Why not a docstring? because a 400 line docstring isn't useful

**Comment by TheCleric** [comment]
> +1 to defining the schema of data frames somewhere. I'd suggest doing it in code rather than text files though. maybe use something like `pandera`, which defines schemas for pandas data frames (and does validation)? https://github.com/pandera-dev/pandera

@schorrm in light of the direction we are thinking about on #155 involving data schemas, I think it would be worth revisiting @bdilday's suggestion above. Especially pandera's ability to handle type coercions:

https://pandera.readthedocs.io/en/stable/dataframe_schemas.html#coercing-types-on-columns

**Comment by TheCleric** [comment]
Although this sounds like a dealbreaker for us:

> Note the special case of integers columns not supporting nan values. In this case, schema.validate will complain if coerce == True and null values are allowed in the column.
> ...
> The best way to handle this case is to simply specify the column as a Float or Object.

So basically it doesn't support nullable int columns. Any nullable int columns would need to be declared as floats.

If that's a dealbreaker we could (ugh) roll our own. For example we could define the schemas as a dict like so:

```python
statcast_schema = {
    'pitch_type': {
        'dtype': 'str',
        'nullable': False,
    },
    'game_date': {
        'dtype': 'date',
        'nullable': False,
        'format': '%Y-%m-%d'
    },
    ...
}
```

We could then use this to feed into something like the `dtypes` and `parse_dates` arguments of `read_csv` or just a generic validator that runs `astype` or targeted `pd.to_numeric` and `pd.to_datetime` calls.

As much as I'm hesitant to roll our own it would solve a problem we have where we have columns that would hard to account for in other schema tools: percentages. I don't see where Pandera supports those and we do some special logic on some where we do string manipulation and division to turn something like `'70.5 %'` to `0.705`.

**Comment by schorrm** [comment]
No way are we rolling our own schema system.

**Comment by TheCleric** [comment]
> No way are we rolling our own schema system.

Completely understand. In that case do we try to figure out how out make Pandera work for us? Or punt this for now until a better solution comes around?

**Comment by schorrm** [comment]
I have yet to have heard any halfway compelling argument for why we need this in the first place, so punting.

**Comment by schorrm** [comment]
even the validation checks could just be a line by line `astype`

**Comment by schorrm** [comment]
Punting. For something that isn't really a problem, it has to make my life easier.

**Comment by schorrm** [comment]
Any objection to closing this issue for now and rebooting this when something changes / becomes more friendly?
@TheCleric  @Am3ra @bdilday

**Comment by TheCleric** [comment]
I'd agree, I'd love to have something, but we don't have a PRESSING need for it, and no clear way forward either.

**Comment by Am3ra** [comment]
Yeah I agree, there's no pressing need.
